### ISSUE TITLE
RNN computation graph does not include cloned RNN parameters

### ISSUE THREAD

**Issue by farid-fari** [learnables/learn2learn#139]
So I have a model with an RNN and a linear layer afterwards:

    class BedPredictor(nn.Module):
        def __init__(self):
            super().__init__()
            self.RNN = nn.LSTM(NUM_VARIABLES, HIDDEN_DIM)
            self.hidden2Pred = nn.Linear(HIDDEN_DIM, NUM_VARIABLES)

        def forward(self, x, hidden=None):
            x = x.view(x.shape[0], -1, NUM_VARIABLES)

            if hidden is None:
                out, hidden = self.RNN(x)
            else:
                out, hidden = self.RNN(x, hidden)

            out = out.squeeze(1)
            pred = self.hidden2Pred(out)

            return pred, hidden

It learns perfectly well without meta-learning on my data, but as soon as I include `learn2learn` it says that there are unused tensors when adapting the learner. Here's my training loop:

    for task in tasks:
            learner = maml.clone()
            learner.train()

            for _ in range(ADAPTATION_STEPS):
                seq = get_task(task)

                prediction, hidden = learner(seq)

                loss = criterion(prediction[:-1], seq[1:])
                learner.adapt(loss)

I've simplified it, removing parts I'm confident the bug does not come from. This makes the program crash unless I specify `allow_unused=True` when instanciating `MAML`, but I should've have to do this.

I decided to investigate the computation graph. The unused tensors from `learner` in the graph are all of the LSTM tensors: `module.RNN.weight_ih_l0`, `module.RNN.weight_hh_l0`, `module.RNN.bias_ih_l0`, `module.RNN.bias_hh_l0`.

I also plotted the backwards graphs for both the normal approach (no meta-learning) and with `l2l` using `torchviz`:

![image](https://user-images.githubusercontent.com/25879088/80087189-12136300-855b-11ea-9bc0-41d5f6e44217.png)

![image](https://user-images.githubusercontent.com/25879088/80087230-1e97bb80-855b-11ea-8aa3-4c730873ae1e.png)

The `Variable`s are in blue, and display their name if they're from either the `maml`, `learner` or original model used to define the MAML. The second graph, using `l2l`, uses RNN parameters which aren't named, and therefore come from somehwere else, though I have no idea how to trace them.

Therefore I think there is an issue with `MAML.clone()` when dealing with RNNs, probably due to the way RNNs work on the inside. This issue persists with `nn.GRU` and `nn.RNN`.

**Comment by farid-fari** [comment]
Upon further testing, this is probably linked to GPU/CUDA. The error about unused gradients persists when using the CPU, however you get the whole computation graph (instead of `CudnnRnnBackward, you get every LSTM operation). The parameters used for this computation are however still somewhere else than `learner`, `maml` and the original model:

![image](https://user-images.githubusercontent.com/25879088/80089624-83084a00-855e-11ea-8130-2fc1c3216a8b.png)

**Comment by seba-1511** [comment]
Thanks for opening the issue @farid-fari . Since the MAML wrapper wasn't extensively tested with RNNs, I'd like to make sure I understand the setup:

1. If you use `allow_unused=True` then everything runs without errors, right ? But in this case the problem is that some of your parameters are not meta-learned.
2. In your code snippet, you call `prediction, hidden = learner(seq)` but not `prediction, hidden = learner(seq, hidden)`. Is that on purpose ?
3. Do you happen to have a minimal example I can use to debug that ? Even a simple Google Colab with fake data that demonstrate the issue would be really helpful.

**Comment by farid-fari** [comment]
1. Correct
2. Yes, pytorch initializes the intial hidden vector to zeroes if you don't provide any.
3. It's generally hard to create minimal examples in meta-learning, but [here it is](https://gist.github.com/farid-fari/8cd4ce8b026af2e6095f1db483a107e5). I first run it without meta-learning and it learns, and then with and it fails. The comments explain where things go wrong and examples without RNNs where it may work.

**Comment by seba-1511** [comment]
Thanks a lot for the minimal example @farid-fari.

The bug is specific to RNN modules; essentially, we need to reset the `self._flat_weights` attribute of the RNN since those weights are used in the forward pass and not the weights in`self._parameters`. For example, such a reset is computed [whenever `self._apply()` is called](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/rnn.py#L148).

I'll push a fix shortly, and cut a new minor release as this is an important issue. In the meantime, you should be able to fix your example by calling:

~~~python
learner.module._apply(lambda x: x)
~~~

before your forward pass. See [this Colab](https://colab.research.google.com/drive/1dpQ6h0gDmIwLF9Duqw9XMRiPHF3cD3Ce) for an example. If you are on master, you can even move that line after `maml.clone()` thanks to the newer `clone_module` implementation.

**Comment by farid-fari** [comment]
Thanks for the very quick fix, impressive! Those weights in `_flat_weights` were indeed well hidden.

**Comment by farid-fari** [comment]
Leaving this also here in case anyone runs across the same issue: when using the CudNN backend, double differentiation is not supported. It's an issue with PyTorch (stemming from CudNN limitations), and has an open bug report: https://github.com/pytorch/pytorch/issues/5261.

The fix is to disable the backend, which seems to have a significant performance hit (but it still runs on GPU):

    torch.backends.cudnn.enabled = False
